include/http/middleware/prerender.js

```diff
@@ -0,0 +1,204 @@
+
+            const option = {
+                userAgent: req.headers['user-agent'],
+                bufferAgent: req.headers['x-bufferbot'],
+                method: req.method,
+                url: req.url
+            };
+
+            if (!shouldPreRender(option)) return;
+
+            const content = await getPrerenderedPage(req);
+            await renderResponse(req, content);
+
+        } catch(ex) {
+            // Just go normally path if there is any errors happen.
+            console.log(ex);
+        }
+    }
+});
+
+const extensionsToIgnore = [
+    '.js',
+    '.css',
+    '.xml',
+    '.less',
+    '.png',
+    '.jpg',
+    '.jpeg',
+    '.gif',
+    '.pdf',
+    '.doc',
+    '.txt',
+    '.ico',
+    '.rss',
+    '.zip',
+    '.mp3',
+    '.rar',
+    '.exe',
+    '.wmv',
+    '.doc',
+    '.avi',
+    '.ppt',
+    '.mpg',
+    '.mpeg',
+    '.tif',
+    '.wav',
+    '.mov',
+    '.psd',
+    '.ai',
+    '.xls',
+    '.mp4',
+    '.m4a',
+    '.swf',
+    '.dat',
+    '.dmg',
+    '.iso',
+    '.flv',
+    '.m4v',
+    '.torrent'
+];
+
+const crawlerUserAgents = [
+    'googlebot',
+    'Yahoo! Slurp',
+    'bingbot',
+    'yandex',
+    'baiduspider',
+    'facebookexternalhit',
+    'twitterbot',
+    'rogerbot',
+    'linkedinbot',
+    'embedly',
+    'quora link preview',
+    'showyoubot',
+    'outbrain',
+    'pinterest/0.',
+    'developers.google.com/+/web/snippet',
+    'slackbot',
+    'vkShare',
+    'W3C_Validator',
+    'redditbot',
+    'Applebot',
+    'WhatsApp',
+    'flipboard',
+    'tumblr',
+    'bitlybot',
+    'SkypeUriPreview',
+    'nuzzel',
+    'Discordbot',
+    'Google Page Speed',
+    'Qwantify',
+    'pinterestbot',
+    'Bitrix link preview',
+    'XING-contenttabreceiver',
+    'Chrome-Lighthouse'
+];
+
+function shouldPreRender(options) {
+    var hasExtensionToIgnore = extensionsToIgnore.some(function (extension) {
+        return options.url.indexOf(extension) !== -1;
+    });
+
+    var isBot = crawlerUserAgents.some(function (crawlerUserAgent) {
+        return options.userAgent.toLowerCase().indexOf(crawlerUserAgent.toLowerCase()) !== -1;
+    });
+
+    // do not pre-rend when:
+    if (!options.userAgent) {
+        return false;
+    }
+
+    if (options.method !== 'GET') {
+        return false;
+    }
+
+    if (hasExtensionToIgnore) {
+        return false;
+    }
+
+    // do pre-render when:
+    var query = url.parse(options.url, true).query;
+    if (query && query['_escaped_fragment_']) {
+        return true;
+    }
+
+    if (options.bufferAgent) {
+        return true;
+    }
+
+    return isBot;
+}
+
+function getPrerenderedPage(req) {
+    var options = {
+        uri: url.parse(buildApiUrl(req)),
+        followRedirect: false,
+        headers: {}
+    };
+
+    Object.keys(req.headers).forEach(function (h) {
+        // Forwarding the host header can cause issues with server platforms that require it to match the URL
+        if (h == 'host') {
```

Comment:
I see '!==' coming quite often, so I believe here you mean to use '===' instead of '=='

Reply:
This is the code copied from third party, I will not modify too much.